dask/order.py

```diff
@@ -202,125 +193,212 @@ def dependencies_key(x):
             StrComparable(x),
         )
 
-    result = {}
-    i = 0
+    def finish_now_key(x):
+        return (-len(dependencies[x]), StrComparable(x))
 
-    # Nodes in the current weakly connected subgraph that are not part of the current
-    # DFS along dependencies.  The next DFS will begin from these nodes.  `outer_stack`
-    # is populated from `init_stack` and the dependents of `outer_stack_seeds`.
-    outer_stack = []
+    partition_keys = {
+        key: (min_dependencies - total_dependencies[key])
+        * (total_dependents - min_heights)
+        for key, (
+            total_dependents,
+            min_dependencies,
+            _,
+            min_heights,
+            _,
+        ) in metrics.items()
+    }
 
-    # Nodes get added to this as they complete.  We will populate `outer_stack` from
-    # dependents of these nodes.  We consider nodes using a FIFO policy (yes, we could
-    # have used collections.deque, but a list with the current index is faster).
-    # Note that this does not include nodes with no dependents!
-    outer_stack_seeds = []
-    outer_stack_seeds_index = 0
+    result = {}
+    i = 0
 
     # Used to perform a DFS along dependencies.  Once emptied (when traversing dependencies),
     # this continues (along dependents) along a path from the initial (leaf) node down to a
     # root node (a tactical goal).  This ensures this root node will get calculated before
     # we consider any other dependents from `outer_stack_seeds`.
-    inner_stack = []
+    inner_stacks = [[min(init_stack, key=initial_stack_key)]]
+    inner_stacks_append = inner_stacks.append
+    inner_stacks_extend = inner_stacks.extend
+    inner_stacks_pop = inner_stacks.pop
 
-    # aliases for speed
-    outer_stack_append = outer_stack.append
-    outer_stack_pop = outer_stack.pop
+    outer_stack = []
     outer_stack_extend = outer_stack.extend
-    outer_stack_seeds_append = outer_stack_seeds.append
-    inner_stack_append = inner_stack.append
-    inner_stack_pop = inner_stack.pop
-    inner_stack_extend = inner_stack.extend
+    outer_stack_pop = outer_stack.pop
+
+    next_nodes = {}
```

Comment:
Could be worth trying and looking at how it affects the benchmarks.  When I've micro-benchmarked this in the past, `defaultdict(list)` is better when the key is usually in the object, otherwise using a `dict` is better.

Reply:
Done.  A couple benchmarks were better for me.  Most were about the same.  I think using `defaultdict(list)` here is more readable, so thanks for the suggestion.